File: sdks/dotnet/src/Ctx.AgentHistory/McpExchange.cs
using System.Text;
using System.Text.Json.Nodes;

namespace Ctx.AgentHistory;

/// <summary>Status reported by an MCP response.</summary>
public enum McpResponseStatus
{
    Succeeded,
    Failed,
    Cancelled,
    TimedOut,
    Unknown
}

/// <summary>Failure category reported for a failed MCP response.</summary>
public enum McpFailureKind
{
    ToolReported,
    Invocation,
    Unknown
}

/// <summary>Capture state for MCP arguments and response JSON.</summary>
public enum McpJsonCaptureStatus
{
    Present,
    Absent,
    Unavailable,
    Omitted
}

/// <summary>Capture state for normalized MCP response text.</summary>
public enum McpTextCaptureStatus
{
    NormalizedBody,
    Absent,
    Unavailable,
    Omitted
}

/// <summary>Reason a complete MCP value was omitted.</summary>
public enum McpPayloadOmissionReason
{
    SizeLimit
}

/// <summary>Closed, content-governed MCP invocation and/or response capture.</summary>
public sealed record McpExchange
{
    public const int MaxIdentityBytes = 64 * 1024;
    public const ulong MaxSafeInteger = 9_007_199_254_740_991UL;

    private readonly JsonObject _json;

    private McpExchange(JsonObject json)
    {
        McpExchangeWire.RequireAllowed(json, "exchange", "providerCallId", "invocation", "response");
        ProviderCallId = McpExchangeWire.RequiredIdentity(json["providerCallId"], "providerCallId");
        Invocation = json.TryGetPropertyValue("invocation", out var invocation)
            ? McpInvocation.FromJson(invocation)
            : null;
        Response = json.TryGetPropertyValue("response", out var response)
            ? McpResponse.FromJson(response)
            : null;
        if (Invocation is null && Response is null)
        {
            throw McpExchangeWire.Invalid("requires invocation, response, or both");
        }
        _json = JsonHelpers.CloneObject(json);
    }

    public string ProviderCallId { get; }
    public McpInvocation? Invocation { get; }
    public McpResponse? Response { get; }

    public JsonObject ToJsonObject() => JsonHelpers.CloneObject(_json);

    internal static McpExchange FromJson(JsonNode? json) =>
        new(McpExchangeWire.RequiredObject(json, "exchange"));

    internal static JsonObject NormalizeWire(JsonNode? json)
    {
        var exchange = McpExchangeWire.NormalizeClosedObject(
            json,
            "exchange",
            new Dictionary<string, string>(StringComparer.Ordinal)
            {
                ["provider_call_id"] = "providerCallId",
                ["providerCallId"] = "providerCallId",
                ["invocation"] = "invocation",
                ["response"] = "response"
            });
        if (exchange.TryGetPropertyValue("invocation", out var invocation))
        {
            exchange["invocation"] = McpInvocation.NormalizeWire(invocation);
        }
        if (exchange.TryGetPropertyValue("response", out var response))
        {
            exchange["response"] = McpResponse.NormalizeWire(response);
        }
        return new McpExchange(exchange).ToJsonObject();
    }
}

/// <summary>Typed MCP invocation metadata and captured arguments.</summary>
public sealed record McpInvocation
{
    private readonly JsonObject _json;

    private McpInvocation(JsonObject json)
    {
        McpExchangeWire.RequireExactly(json, "invocation", "server", "tool", "arguments");
        Server = McpExchangeWire.RequiredIdentity(json["server"], "invocation.server");
        Tool = McpExchangeWire.RequiredIdentity(json["tool"], "invocation.tool");
        Arguments = McpArgumentsCapture.FromJson(json["arguments"]);
        _json = JsonHelpers.CloneObject(json);
    }

    public string Server { get; }
    public string Tool { get; }
    public McpArgumentsCapture Arguments { get; }

    public JsonObject ToJsonObject() => JsonHelpers.CloneObject(_json);

    internal static McpInvocation FromJson(JsonNode? json) =>
        new(McpExchangeWire.RequiredObject(json, "invocation"));

    internal static JsonObject NormalizeWire(JsonNode? json)
    {
        var invocation = McpExchangeWire.NormalizeClosedObject(
            json,
            "invocation",
            new Dictionary<string, string>(StringComparer.Ordinal)
            {
                ["server"] = "server",
                ["tool"] = "tool",
                ["arguments"] = "arguments"
            });
        if (invocation.TryGetPropertyValue("arguments", out var arguments))
        {
            invocation["arguments"] = McpArgumentsCapture.NormalizeWire(arguments);
        }
        return new McpInvocation(invocation).ToJsonObject();
    }
}

/// <summary>Typed MCP response metadata and capture dispositions.</summary>
public sealed record McpResponse
{
    private readonly JsonObject _json;

    private McpResponse(JsonObject json)
    {
        McpExchangeWire.RequireAllowed(
            json,
            "response",
            "status",
            "failureKind",
            "durationNs",
            "text",
            "payload");
        McpExchangeWire.RequireMembers(json, "response", "status", "text", "payload");

        Status = McpExchangeWire.ResponseStatus(json["status"]);
        FailureKind = json.TryGetPropertyValue("failureKind", out var failureKind)
            ? McpExchangeWire.FailureKind(failureKind)
            : null;
        if ((Status == McpResponseStatus.Failed) != (FailureKind is not null))
        {
            throw McpExchangeWire.Invalid("failureKind must be present exactly for failed responses");
        }
        DurationNs = json.TryGetPropertyValue("durationNs", out var durationNs)
            ? McpExchangeWire.SafeInteger(durationNs, "response.durationNs")
            : null;
        Text = McpTextCapture.FromJson(json["text"]);
        Payload = McpJsonCapture.FromJson(json["payload"], "response.payload");
        _json = JsonHelpers.CloneObject(json);
    }

    public McpResponseStatus Status { get; }
    public McpFailureKind? FailureKind { get; }
    public ulong? DurationNs { get; }
    public McpTextCapture Text { get; }
    public McpJsonCapture Payload { get; }

    public JsonObject ToJsonObject() => JsonHelpers.CloneObject(_json);

    internal static McpResponse FromJson(JsonNode? json) =>
        new(McpExchangeWire.RequiredObject(json, "response"));

    internal static JsonObject NormalizeWire(JsonNode? json)
    {
        var response = McpExchangeWire.NormalizeClosedObject(
            json,
            "response",
            new Dictionary<string, string>(StringComparer.Ordinal)
            {
                ["status"] = "status",
                ["failure_kind"] = "failureKind",
                ["failureKind"] = "failureKind",
                ["duration_ns"] = "durationNs",
                ["durationNs"] = "durationNs",
                ["text"] = "text",
                ["payload"] = "payload"
            });
        if (response.TryGetPropertyValue("text", out var text))
        {
            response["text"] = McpTextCapture.NormalizeWire(text);
        }
        if (response.TryGetPropertyValue("payload", out var payload))
        {
            response["payload"] = McpJsonCapture.NormalizeWire(payload, "response.payload");
        }
        return new McpResponse(response).ToJsonObject();
    }
}

/// <summary>MCP invocation-argument capture. Present values are JSON objects.</summary>
public sealed record McpArgumentsCapture
{
    private readonly JsonObject _json;

    private McpArgumentsCapture(JsonObject json)
    {
        CaptureStatus = McpExchangeWire.JsonCaptureStatus(json["captureStatus"], "invocation.arguments");
        switch (CaptureStatus)
        {
            case McpJsonCaptureStatus.Present:
                McpExchangeWire.RequireExactly(json, "invocation.arguments", "captureStatus", "value");
                Value = json["value"] is JsonObject value
                    ? JsonHelpers.CloneObject(value)
                    : throw McpExchangeWire.Invalid(
                        "present invocation arguments must be a JSON object",
                        "invocation.arguments.value");
                break;
            case McpJsonCaptureStatus.Omitted:
                (Reason, ObservedEncodedBytes) = McpExchangeWire.ReadOmission(json, "invocation.arguments");
                break;
            default:
                McpExchangeWire.RequireExactly(json, "invocation.arguments", "captureStatus");
                break;
        }
        _json = JsonHelpers.CloneObject(json);
    }

    public McpJsonCaptureStatus CaptureStatus { get; }
    public JsonObject? Value { get; }
    public McpPayloadOmissionReason? Reason { get; }
    public ulong? ObservedEncodedBytes { get; }

    public JsonObject ToJsonObject() => JsonHelpers.CloneObject(_json);

    internal static McpArgumentsCapture FromJson(JsonNode? json) =>
        new(McpExchangeWire.RequiredObject(json, "invocation.arguments"));

    internal static JsonObject NormalizeWire(JsonNode? json)
    {
        var capture = McpExchangeWire.NormalizeCaptureObject(json, "invocation.arguments");
        return new McpArgumentsCapture(capture).ToJsonObject();
    }
}

/// <summary>MCP JSON capture. Present values preserve arbitrary JSON keys and nulls.</summary>
public sealed record McpJsonCapture
{
    private readonly JsonObject _json;

    private McpJsonCapture(JsonObject json, string context)
    {
        CaptureStatus = McpExchangeWire.JsonCaptureStatus(json["captureStatus"], context);
        switch (CaptureStatus)
        {
            case McpJsonCaptureStatus.Present:
                McpExchangeWire.RequireExactly(json, context, "captureStatus", "value");
                Value = JsonHelpers.Clone(json["value"]);
                break;
            case McpJsonCaptureStatus.Omitted:
                (Reason, ObservedEncodedBytes) = McpExchangeWire.ReadOmission(json, context);
                break;
            default:
                McpExchangeWire.RequireExactly(json, context, "captureStatus");
                break;
        }
        _json = JsonHelpers.CloneObject(json);
    }

    public McpJsonCaptureStatus CaptureStatus { get; }
    public JsonNode? Value { get; }
    public McpPayloadOmissionReason? Reason { get; }
    public ulong? ObservedEncodedBytes { get; }

    public JsonObject ToJsonObject() => JsonHelpers.CloneObject(_json);

    internal static McpJsonCapture FromJson(JsonNode? json, string context) =>
        new(McpExchangeWire.RequiredObject(json, context), context);

    internal static JsonObject NormalizeWire(JsonNode? json, string context)
    {
        var capture = McpExchangeWire.NormalizeCaptureObject(json, context);
        return new McpJsonCapture(capture, context).ToJsonObject();
    }
}

/// <summary>MCP response-text capture disposition.</summary>
public sealed record McpTextCapture
{
    private readonly JsonObject _json;

    private McpTextCapture(JsonObject json)
    {
        CaptureStatus = McpExchangeWire.TextCaptureStatus(json["captureStatus"]);
        if (CaptureStatus == McpTextCaptureStatus.Omitted)
        {
            (Reason, ObservedEncodedBytes) = McpExchangeWire.ReadOmission(json, "response.text");
        }
        else
        {
            McpExchangeWire.RequireExactly(json, "response.text", "captureStatus");
        }
        _json = JsonHelpers.CloneObject(json);
    }

    public McpTextCaptureStatus CaptureStatus { get; }
    public McpPayloadOmissionReason? Reason { get; }
    public ulong? ObservedEncodedBytes { get; }

    public JsonObject ToJsonObject() => JsonHelpers.CloneObject(_json);

    internal static McpTextCapture FromJson(JsonNode? json) =>
        new(McpExchangeWire.RequiredObject(json, "response.text"));

    internal static JsonObject NormalizeWire(JsonNode? json)
    {
        var capture = McpExchangeWire.NormalizeCaptureObject(json, "response.text");
        return new McpTextCapture(capture).ToJsonObject();
    }
}

internal static class McpExchangeWire
{
    private static readonly UTF8Encoding StrictUtf8 = new(false, true);

    public static JsonObject NormalizeCaptureObject(JsonNode? json, string context) =>
        NormalizeClosedObject(
            json,
            context,
            new Dictionary<string, string>(StringComparer.Ordinal)
            {
                ["capture_status"] = "captureStatus",
                ["captureStatus"] = "captureStatus",
                ["value"] = "value",
                ["reason"] = "reason",
                ["observed_encoded_bytes"] = "observedEncodedBytes",
                ["observedEncodedBytes"] = "observedEncodedBytes"
            });

    public static JsonObject NormalizeClosedObject(
        JsonNode? json,
        string context,
        IReadOnlyDictionary<string, string> aliases)
    {
        var input = RequiredObject(json, context);
        var result = new JsonObject();
        foreach (var pair in input)
        {
            if (!aliases.TryGetValue(pair.Key, out var canonical))
            {
                throw Invalid($"{context} contains unknown member {pair.Key}", context);
            }
            if (result.ContainsKey(canonical))
            {
                throw Invalid($"{context} contains colliding aliases for {canonical}", context);
            }
            result[canonical] = JsonHelpers.Clone(pair.Value);
        }
        return result;
    }

    public static JsonObject RequiredObject(JsonNode? json, string context) =>
        json is JsonObject value
            ? value
            : throw Invalid($"{context} must be an object", context);

    public static void RequireExactly(JsonObject json, string context, params string[] fields)
    {
        RequireAllowed(json, context, fields);
        RequireMembers(json, context, fields);
    }

    public static void RequireAllowed(JsonObject json, string context, params string[] fields)
    {
        var allowed = new HashSet<string>(fields, StringComparer.Ordinal);
        var unknown = json.Select(pair => pair.Key).Where(key => !allowed.Contains(key)).ToArray();
        if (unknown.Length > 0)
        {
            throw Invalid($"{context} contains unknown members: {string.Join(", ", unknown)}", context);
        }
    }

    public static void RequireMembers(JsonObject json, string context, params string[] fields)
    {
        foreach (var field in fields)
        {
            if (!json.ContainsKey(field))
            {
                throw Invalid($"{context} requires {field}", $"{context}.{field}");
            }
        }
    }

    public static string RequiredIdentity(JsonNode? json, string field)
    {
        if (json is not JsonValue value || !value.TryGetValue<string>(out var text) || text.Length == 0)
        {
            throw Invalid($"{field} must be a nonempty string", field);
        }
        try
        {
            if (StrictUtf8.GetByteCount(text) > McpExchange.MaxIdentityBytes)
            {
                throw Invalid(
                    $"{field} exceeds {McpExchange.MaxIdentityBytes} decoded UTF-8 bytes",
                    field);
            }
        }
        catch (EncoderFallbackException exception)
        {
            throw Invalid($"{field} contains an invalid Unicode string", field, exception);
        }
        return text;
    }

    public static ulong SafeInteger(JsonNode? json, string field)
    {
        if (json is not JsonValue value
            || !value.TryGetValue<ulong>(out var number)
            || number > McpExchange.MaxSafeInteger)
        {
            throw Invalid(
                $"{field} is outside the exact JSON integer domain",
                field);
        }
        return number;
    }

    public static McpResponseStatus ResponseStatus(JsonNode? json) =>
        RequiredWireString(json, "response.status") switch
        {
            "succeeded" => McpResponseStatus.Succeeded,
            "failed" => McpResponseStatus.Failed,
            "cancelled" => McpResponseStatus.Cancelled,
            "timed_out" => McpResponseStatus.TimedOut,
            "unknown" => McpResponseStatus.Unknown,
            _ => throw Invalid("response.status is invalid", "response.status")
        };

    public static McpFailureKind FailureKind(JsonNode? json) =>
        RequiredWireString(json, "response.failureKind") switch
        {
            "tool_reported" => McpFailureKind.ToolReported,
            "invocation" => McpFailureKind.Invocation,
            "unknown" => McpFailureKind.Unknown,
            _ => throw Invalid("response.failureKind is invalid", "response.failureKind")
        };

    public static McpJsonCaptureStatus JsonCaptureStatus(JsonNode? json, string context) =>
        RequiredWireString(json, $"{context}.captureStatus") switch
        {
            "present" => McpJsonCaptureStatus.Present,
            "absent" => McpJsonCaptureStatus.Absent,
            "unavailable" => McpJsonCaptureStatus.Unavailable,
            "omitted" => McpJsonCaptureStatus.Omitted,
            _ => throw Invalid($"{context}.captureStatus is invalid", $"{context}.captureStatus")
        };

    public static McpTextCaptureStatus TextCaptureStatus(JsonNode? json) =>
        RequiredWireString(json, "response.text.captureStatus") switch
        {
            "normalized_body" => McpTextCaptureStatus.NormalizedBody,
            "absent" => McpTextCaptureStatus.Absent,
            "unavailable" => McpTextCaptureStatus.Unavailable,
            "omitted" => McpTextCaptureStatus.Omitted,
            _ => throw Invalid("response.text.captureStatus is invalid", "response.text.captureStatus")
        };

    public static (McpPayloadOmissionReason Reason, ulong? ObservedEncodedBytes) ReadOmission(
        JsonObject json,
        string context)
    {
        RequireAllowed(json, context, "captureStatus", "reason", "observedEncodedBytes");
        RequireMembers(json, context, "captureStatus", "reason");
        var reason = RequiredWireString(json["reason"], $"{context}.reason") switch
        {
            "size_limit" => McpPayloadOmissionReason.SizeLimit,
            _ => throw Invalid($"{context}.reason is invalid", $"{context}.reason")
        };
        ulong? observed = json.TryGetPropertyValue("observedEncodedBytes", out var value)
            ? SafeInteger(value, $"{context}.observedEncodedBytes")
            : null;
        return (reason, observed);
    }

    private static string RequiredWireString(JsonNode? json, string field) =>
        json is JsonValue value && value.TryGetValue<string>(out var text)
            ? text
            : throw Invalid($"{field} must be a string", field);

    public static CtxAgentHistoryProtocolException Invalid(
        string message,
        string? field = null,
        Exception? exception = null) =>
        new(
            $"agent-history-v1 MCP exchange {message}",
            ErrorDetails(message, field),
            exception);

    private static JsonObject ErrorDetails(string message, string? field)
    {
        var details = new JsonObject
        {
            ["field"] = field is null ? "mcpExchange" : $"mcpExchange.{field}"
        };
        if (message.Contains("exact JSON integer domain", StringComparison.Ordinal))
        {
            details["maximum"] = McpExchange.MaxSafeInteger;
        }
        return details;
    }
}

File: sdks/jvm/src/main/java/rs/ctx/agenthistory/McpExchange.java
package rs.ctx.agenthistory;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.nio.CharBuffer;
import java.nio.charset.CharacterCodingException;
import java.nio.charset.CodingErrorAction;
import java.nio.charset.StandardCharsets;
import java.util.LinkedHashMap;
import java.util.Map;

/** Content-governed MCP invocation and response capture. */
public final class McpExchange {
    public static final int MAX_IDENTITY_BYTES = 64 * 1024;
    public static final long MAX_SAFE_INTEGER = 9_007_199_254_740_991L;

    private final Map<String, Object> fields;
    private final String providerCallId;
    private final Invocation invocation;
    private final Response response;

    McpExchange(Map<String, Object> fields) {
        requireAllowed(fields, "exchange", "providerCallId", "invocation", "response");
        this.providerCallId = requiredIdentity(fields.get("providerCallId"), "providerCallId");
        this.invocation = fields.containsKey("invocation")
                ? Invocation.from(fields.get("invocation"))
                : null;
        this.response = fields.containsKey("response")
                ? Response.from(fields.get("response"))
                : null;
        if (invocation == null && response == null) {
            throw protocolError("requires invocation, response, or both", null);
        }
        this.fields = AgentHistoryValue.copyObject(fields);
    }

    static McpExchange from(Object value) {
        Map<String, Object> fields = AgentHistoryValue.objectOrNull(value);
        if (fields == null) {
            throw protocolError("must be an object when present", null);
        }
        return new McpExchange(fields);
    }

    static Map<String, Object> normalizeWire(Object value) {
        Map<String, Object> exchange = normalizeObject(
                value,
                "exchange",
                aliases(
                        "provider_call_id", "providerCallId",
                        "providerCallId", "providerCallId",
                        "invocation", "invocation",
                        "response", "response"));
        if (exchange.containsKey("invocation")) {
            exchange.put("invocation", normalizeInvocation(exchange.get("invocation")));
        }
        if (exchange.containsKey("response")) {
            exchange.put("response", normalizeResponse(exchange.get("response")));
        }
        return new McpExchange(exchange).asMap();
    }

    public String getProviderCallId() {
        return providerCallId;
    }

    public String providerCallId() {
        return providerCallId;
    }

    public Invocation getInvocation() {
        return invocation;
    }

    public Invocation invocation() {
        return invocation;
    }

    public Response getResponse() {
        return response;
    }

    public Response response() {
        return response;
    }

    public Map<String, Object> asMap() {
        return fields;
    }

    public enum ResponseStatus {
        SUCCEEDED("succeeded"),
        FAILED("failed"),
        CANCELLED("cancelled"),
        TIMED_OUT("timed_out"),
        UNKNOWN("unknown");

        private final String wireValue;

        ResponseStatus(String wireValue) {
            this.wireValue = wireValue;
        }

        public String wireValue() {
            return wireValue;
        }
    }

    public enum FailureKind {
        TOOL_REPORTED("tool_reported"),
        INVOCATION("invocation"),
        UNKNOWN("unknown");

        private final String wireValue;

        FailureKind(String wireValue) {
            this.wireValue = wireValue;
        }

        public String wireValue() {
            return wireValue;
        }
    }

    public enum JsonCaptureStatus {
        PRESENT("present"),
        ABSENT("absent"),
        UNAVAILABLE("unavailable"),
        OMITTED("omitted");

        private final String wireValue;

        JsonCaptureStatus(String wireValue) {
            this.wireValue = wireValue;
        }

        public String wireValue() {
            return wireValue;
        }
    }

    public enum TextCaptureStatus {
        NORMALIZED_BODY("normalized_body"),
        ABSENT("absent"),
        UNAVAILABLE("unavailable"),
        OMITTED("omitted");

        private final String wireValue;

        TextCaptureStatus(String wireValue) {
            this.wireValue = wireValue;
        }

        public String wireValue() {
            return wireValue;
        }
    }

    public enum OmissionReason {
        SIZE_LIMIT("size_limit");

        private final String wireValue;

        OmissionReason(String wireValue) {
            this.wireValue = wireValue;
        }

        public String wireValue() {
            return wireValue;
        }
    }

    public static final class Invocation {
        private final Map<String, Object> fields;
        private final String server;
        private final String tool;
        private final JsonCapture arguments;

        private Invocation(Map<String, Object> fields) {
            requireExactly(fields, "invocation", "server", "tool", "arguments");
            server = requiredIdentity(fields.get("server"), "invocation.server");
            tool = requiredIdentity(fields.get("tool"), "invocation.tool");
            arguments = JsonCapture.from(fields.get("arguments"), "invocation.arguments");
            if (arguments.status == JsonCaptureStatus.PRESENT
                    && !(arguments.value instanceof Map<?, ?>)) {
                throw protocolError("present invocation arguments must be a JSON object", null);
            }
            this.fields = AgentHistoryValue.copyObject(fields);
        }

        static Invocation from(Object value) {
            return new Invocation(requiredObject(value, "invocation"));
        }

        public String getServer() { return server; }
        public String server() { return server; }
        public String getTool() { return tool; }
        public String tool() { return tool; }
        public JsonCapture getArguments() { return arguments; }
        public JsonCapture arguments() { return arguments; }
        public Map<String, Object> asMap() { return fields; }
    }

    public static final class Response {
        private final Map<String, Object> fields;
        private final ResponseStatus status;
        private final FailureKind failureKind;
        private final Long durationNs;
        private final TextCapture text;
        private final JsonCapture payload;

        private Response(Map<String, Object> fields) {
            requireAllowed(fields, "response", "status", "failureKind", "durationNs", "text", "payload");
            requireMembers(fields, "response", "status", "text", "payload");
            status = enumValue(ResponseStatus.values(), fields.get("status"), "response.status");
            failureKind = fields.containsKey("failureKind")
                    ? enumValue(FailureKind.values(), fields.get("failureKind"), "response.failureKind")
                    : null;
            if ((status == ResponseStatus.FAILED) != (failureKind != null)) {
                throw protocolError("failureKind must be present exactly for failed responses", null);
            }
            durationNs = fields.containsKey("durationNs")
                    ? safeInteger(fields.get("durationNs"), "response.durationNs")
                    : null;
            text = TextCapture.from(fields.get("text"));
            payload = JsonCapture.from(fields.get("payload"), "response.payload");
            this.fields = AgentHistoryValue.copyObject(fields);
        }

        static Response from(Object value) {
            return new Response(requiredObject(value, "response"));
        }

        public ResponseStatus getStatus() { return status; }
        public ResponseStatus status() { return status; }
        public FailureKind getFailureKind() { return failureKind; }
        public FailureKind failureKind() { return failureKind; }
        public Long getDurationNs() { return durationNs; }
        public Long durationNs() { return durationNs; }
        public TextCapture getText() { return text; }
        public TextCapture text() { return text; }
        public JsonCapture getPayload() { return payload; }
        public JsonCapture payload() { return payload; }
        public Map<String, Object> asMap() { return fields; }
    }

    public static final class JsonCapture {
        private final Map<String, Object> fields;
        private final JsonCaptureStatus status;
        private final Object value;
        private final OmissionReason reason;
        private final Long observedEncodedBytes;

        private JsonCapture(Map<String, Object> fields, String context) {
            status = enumValue(JsonCaptureStatus.values(), fields.get("captureStatus"), context + ".captureStatus");
            if (status == JsonCaptureStatus.PRESENT) {
                requireExactly(fields, context, "captureStatus", "value");
                value = AgentHistoryValue.copy(fields.get("value"));
                reason = null;
                observedEncodedBytes = null;
            } else if (status == JsonCaptureStatus.OMITTED) {
                requireAllowed(fields, context, "captureStatus", "reason", "observedEncodedBytes");
                requireMembers(fields, context, "captureStatus", "reason");
                value = null;
                reason = enumValue(OmissionReason.values(), fields.get("reason"), context + ".reason");
                observedEncodedBytes = fields.containsKey("observedEncodedBytes")
                        ? safeInteger(fields.get("observedEncodedBytes"), context + ".observedEncodedBytes")
                        : null;
            } else {
                requireExactly(fields, context, "captureStatus");
                value = null;
                reason = null;
                observedEncodedBytes = null;
            }
            this.fields = AgentHistoryValue.copyObject(fields);
        }

        static JsonCapture from(Object value, String context) {
            return new JsonCapture(requiredObject(value, context), context);
        }

        public JsonCaptureStatus getCaptureStatus() { return status; }
        public JsonCaptureStatus captureStatus() { return status; }
        public Object getValue() { return value; }
        public Object value() { return value; }
        public OmissionReason getReason() { return reason; }
        public Long getObservedEncodedBytes() { return observedEncodedBytes; }
        public Map<String, Object> asMap() { return fields; }
    }

    public static final class TextCapture {
        private final Map<String, Object> fields;
        private final TextCaptureStatus status;
        private final OmissionReason reason;
        private final Long observedEncodedBytes;

        private TextCapture(Map<String, Object> fields) {
            status = enumValue(TextCaptureStatus.values(), fields.get("captureStatus"), "response.text.captureStatus");
            if (status == TextCaptureStatus.OMITTED) {
                requireAllowed(fields, "response.text", "captureStatus", "reason", "observedEncodedBytes");
                requireMembers(fields, "response.text", "captureStatus", "reason");
                reason = enumValue(OmissionReason.values(), fields.get("reason"), "response.text.reason");
                observedEncodedBytes = fields.containsKey("observedEncodedBytes")
                        ? safeInteger(fields.get("observedEncodedBytes"), "response.text.observedEncodedBytes")
                        : null;
            } else {
                requireExactly(fields, "response.text", "captureStatus");
                reason = null;
                observedEncodedBytes = null;
            }
            this.fields = AgentHistoryValue.copyObject(fields);
        }

        static TextCapture from(Object value) {
            return new TextCapture(requiredObject(value, "response.text"));
        }

        public TextCaptureStatus getCaptureStatus() { return status; }
        public TextCaptureStatus captureStatus() { return status; }
        public OmissionReason getReason() { return reason; }
        public Long getObservedEncodedBytes() { return observedEncodedBytes; }
        public Map<String, Object> asMap() { return fields; }
    }

    private static Map<String, Object> normalizeInvocation(Object value) {
        Map<String, Object> invocation = normalizeObject(
                value, "invocation", aliases("server", "server", "tool", "tool", "arguments", "arguments"));
        if (invocation.containsKey("arguments")) {
            invocation.put("arguments", normalizeCapture(invocation.get("arguments"), "invocation.arguments", false));
        }
        return invocation;
    }

    private static Map<String, Object> normalizeResponse(Object value) {
        Map<String, Object> response = normalizeObject(
                value,
                "response",
                aliases(
                        "status", "status",
                        "failure_kind", "failureKind",
                        "failureKind", "failureKind",
                        "duration_ns", "durationNs",
                        "durationNs", "durationNs",
                        "text", "text",
                        "payload", "payload"));
        if (response.containsKey("text")) {
            response.put("text", normalizeCapture(response.get("text"), "response.text", true));
        }
        if (response.containsKey("payload")) {
            response.put("payload", normalizeCapture(response.get("payload"), "response.payload", false));
        }
        return response;
    }

    private static Map<String, Object> normalizeCapture(Object value, String context, boolean text) {
        Map<String, Object> capture = normalizeObject(
                value,
                context,
                aliases(
                        "capture_status", "captureStatus",
                        "captureStatus", "captureStatus",
                        "value", "value",
                        "reason", "reason",
                        "observed_encoded_bytes", "observedEncodedBytes",
                        "observedEncodedBytes", "observedEncodedBytes"));
        if (text) {
            new TextCapture(capture);
        } else {
            new JsonCapture(capture, context);
        }
        return capture;
    }

    private static Map<String, Object> normalizeObject(
            Object value, String context, Map<String, String> aliases) {
        Map<String, Object> input = AgentHistoryValue.objectOrNull(value);
        if (input == null) {
            throw protocolError(context + " must be an object", null);
        }
        Map<String, Object> out = new LinkedHashMap<>();
        for (Map.Entry<String, Object> entry : input.entrySet()) {
            String canonical = aliases.get(entry.getKey());
            if (canonical == null) {
                throw protocolError(context + " contains unknown member " + entry.getKey(), null);
            }
            if (out.containsKey(canonical)) {
                throw protocolError(context + " contains colliding aliases for " + canonical, null);
            }
            out.put(canonical, AgentHistoryValue.copy(entry.getValue()));
        }
        return out;
    }

    private static Map<String, String> aliases(String... entries) {
        Map<String, String> aliases = new LinkedHashMap<>();
        for (int index = 0; index < entries.length; index += 2) {
            aliases.put(entries[index], entries[index + 1]);
        }
        return aliases;
    }

    private static Map<String, Object> requiredObject(Object value, String context) {
        Map<String, Object> object = AgentHistoryValue.objectOrNull(value);
        if (object == null) {
            throw protocolError(context + " must be an object", null);
        }
        return object;
    }

    private static void requireExactly(Map<String, Object> value, String context, String... fields) {
        requireAllowed(value, context, fields);
        requireMembers(value, context, fields);
    }

    private static void requireAllowed(Map<String, Object> value, String context, String... fields) {
        java.util.Set<String> allowed = java.util.Set.of(fields);
        for (String field : value.keySet()) {
            if (!allowed.contains(field)) {
                throw protocolError(context + " contains unknown member " + field, null);
            }
        }
    }

    private static void requireMembers(Map<String, Object> value, String context, String... fields) {
        for (String field : fields) {
            if (!value.containsKey(field)) {
                throw protocolError(context + " requires " + field, null);
            }
        }
    }

    private static String requiredIdentity(Object value, String field) {
        if (!(value instanceof String) || ((String) value).isEmpty()) {
            throw protocolError(field + " must be a nonempty string", null);
        }
        String text = (String) value;
        try {
            int bytes = StandardCharsets.UTF_8.newEncoder()
                    .onMalformedInput(CodingErrorAction.REPORT)
                    .onUnmappableCharacter(CodingErrorAction.REPORT)
                    .encode(CharBuffer.wrap(text))
                    .remaining();
            if (bytes > MAX_IDENTITY_BYTES) {
                throw protocolError(field + " exceeds " + MAX_IDENTITY_BYTES + " decoded UTF-8 bytes", null);
            }
        } catch (CharacterCodingException cause) {
            throw protocolError(field + " contains invalid Unicode", cause);
        }
        return text;
    }

    private static Long safeInteger(Object value, String field) {
        BigInteger integer;
        try {
            if (value instanceof BigDecimal) {
                integer = ((BigDecimal) value).toBigIntegerExact();
            } else if (value instanceof Byte || value instanceof Short
                    || value instanceof Integer || value instanceof Long) {
                integer = BigInteger.valueOf(((Number) value).longValue());
            } else {
                throw new ArithmeticException();
            }
        } catch (ArithmeticException ignored) {
            throw protocolError(field + " is outside the exact JSON integer domain", null);
        }
        if (integer.signum() < 0 || integer.compareTo(BigInteger.valueOf(MAX_SAFE_INTEGER)) > 0) {
            throw protocolError(field + " is outside the exact JSON integer domain", null);
        }
        return Long.valueOf(integer.longValue());
    }

    private static <T extends Enum<T>> T enumValue(T[] values, Object value, String field) {
        if (!(value instanceof String)) {
            throw protocolError(field + " must be a string", null);
        }
        for (T candidate : values) {
            if (candidate instanceof ResponseStatus
                    && ((ResponseStatus) candidate).wireValue().equals(value)
                    || candidate instanceof FailureKind
                    && ((FailureKind) candidate).wireValue().equals(value)
                    || candidate instanceof JsonCaptureStatus
                    && ((JsonCaptureStatus) candidate).wireValue().equals(value)
                    || candidate instanceof TextCaptureStatus
                    && ((TextCaptureStatus) candidate).wireValue().equals(value)
                    || candidate instanceof OmissionReason
                    && ((OmissionReason) candidate).wireValue().equals(value)) {
                return candidate;
            }
        }
        throw protocolError(field + " has unknown value " + value, null);
    }

    static CtxAgentHistoryException.Protocol protocolError(String message, Throwable cause) {
        Map<String, Object> details = new LinkedHashMap<>();
        details.put("field", "mcpExchange");
        return new CtxAgentHistoryException.Protocol(
                "agent-history-v1 MCP exchange " + message,
                details,
                cause);
    }
}
